code_generator/GdkCodeGenerator/Templates/UnityCommandPayloadGenerator.tt

```diff
@@ -17,92 +18,87 @@ namespace <#= qualifiedNamespace #>
                 {
-                    this.RequestId = requestId;
-                    this.Translation = translation;
-                    this.CallerWorkerId = callerWorkerId;
-                    this.CallerAttributeSet = callerAttributeSet;
-                    this.RawRequest = request;
-                }
-
-                public void Send<#= commandDetails.CommandName #>Response(<#= commandDetails.ResponseType #> response)
-                {
-                    Translation.<#= commandDetails.CommandName #>Responses.Add(
-                        new OutgoingResponse(RequestId, response));
+                    RequestId = requestId;
+                    CallerWorkerId = callerWorkerId;
+                    CallerAttributeSet = callerAttributeSet;
+                    RawRequest = request;
                 }
             }
 
-            internal struct OutgoingRequest : IOutgoingCommandRequest
+            public struct Response
             {
-                public long TargetEntityId { get; }
+                public uint RequestId { get; }
+                public <#= commandDetails.ResponseType #>? RawResponse { get; }
+                public string FailureMessage { get; }
 
-                public long SenderEntityId { get; }
+                internal Response(ReceivedRequest req, <#= commandDetails.ResponseType #>? payload, string failureMessage)
+                {
+                    RequestId = req.RequestId;
+                    RawResponse = payload;
+                    FailureMessage = failureMessage;
+                }
 
-                public <#= commandDetails.RequestType #> RawRequest { get; }
+                public static Response CreateResponse(ReceivedRequest req, <#= commandDetails.ResponseType #> payload)
+                {
+                    return new Response(req, payload, null);
+                }
 
-                internal OutgoingRequest(long targetEntityId, long senderEntityId,
-                    <#= commandDetails.RequestType #> request)
+                public static Response CreateFailure(ReceivedRequest req, string failureMessage)
                 {
-                    this.TargetEntityId = targetEntityId;
-                    this.SenderEntityId = senderEntityId;
-                    this.RawRequest = request;
+                    return new Response(req, null, failureMessage);
                 }
             }
 
-            public struct Response : IIncomingCommandResponse
+            public struct ReceivedResponse
             {
-                public long EntityId { get; }
-
+                public EntityId EntityId { get; }
                 public string Message { get; }
-
-                public CommandStatusCode StatusCode { get; }
-
+                public StatusCode StatusCode { get; }
                 public <#= commandDetails.ResponseType #>? RawResponse { get; }
-
                 public <#= commandDetails.RequestType #> RawRequest { get; }
```

Comment:
Not strictly part of this PR but RawRequest doesn't make sense here. RequestPayload or SentRequest would work better.

Reply:
I've got a cleanup pass planned - so can factor all this in :)